First attempt to add testing

kyrcha opened the pull request:
Used the `git-fixtures` organization to test the download of a repository and its descendants (PR, PR comments, etc.) and the download of an organization and its descendants (Users etc.).

It uses an online but fixed organization in order to test that the graphql client, the api endpoint and the parsing works as expected.

lwsanty commented on .gitignore:
```diff
@@ -0,0 +1 @@
+*.out
```
![dsfdsfsdfsdfsdaf](https://user-images.githubusercontent.com/10910699/66133637-aa69ea80-e5ff-11e9-97be-c8926e0cc2f6.png)

lwsanty commented on github/downloader_test.go:
```diff
@@ -0,0 +1,205 @@
+		require.Equal(t, test.totalPrivateRepos, org.TotalPrivateRepos.TotalCount, "Organization total private repos count should be the same")
+		require.Equal(t, test.numOfUsers, teststore.GetNumOfUsers(), "Organization total private repos count should be the same")
+	}
+}
```
this tests structure is similar to the previous one so please take into account aforementioned comments

lwsanty commented on github/downloader_test.go:
```diff
@@ -0,0 +1,205 @@
+	"golang.org/x/oauth2"
+)
+
+func TestOnlineRepositoryDownload(t *testing.T) {
```
please add a couple of lines of doc here, [some rules](https://blog.golang.org/godoc-documenting-go-code)

lwsanty commented on github/store/teststore.go:
```diff
@@ -0,0 +1,117 @@
+
+func (s *TestStore) SaveOrganization(organization *graphql.Organization) error {
+	fmt.Printf("organization data fetched for %s\n", organization.Login)
+	// fmt.Printf("%+v\n", organization)
```
please get rid of commented debugs or use logs on debug level [here](https://github.com/src-d/go-log#configuration)

lwsanty commented on github/store/teststore.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+func (s *TestStore) SaveOrganization(organization *graphql.Organization) error {
+	fmt.Printf("organization data fetched for %s\n", organization.Login)
```
according to logging please consider using the logger from our code convention [here](https://github.com/src-d/guide/blob/806661020bf47a2a736e7b72aee1095d42a41774/engineering/conventions/go.md#logging)

lwsanty commented on github/store/teststore.go:
```diff
@@ -0,0 +1,117 @@
+
+func (s *TestStore) GetNumOfUsers() int {
+	return len(s.users)
+}
```
I don't see that these methods would help a lot during the testing because it's still easy to access the fields

smacker commented on github/downloader_test.go:
```diff
@@ -0,0 +1,205 @@
+	"github.com/src-d/metadata-retrieval/github/store"
+	"github.com/stretchr/testify/require"
+
+	"golang.org/x/oauth2"
```
as this repo is maintained by apps team the result of autoformat made by `goimports` is good enough for us.

dpordomingo replied:
I don't see a problem with adopting company-wide rules, but if there's a strong opinion against it, we can postpone the discussion and proceed as @smacker suggests.

dpordomingo replied:
I don't see a problem with adopting company-wide rules, but if there's a strong opinion against it, we can postpone the discussion and proceed as @smacker suggests.

smacker commented on github/downloader_test.go:
```diff
@@ -0,0 +1,205 @@
+			3,
+			1,
+		},
+	}
```
the dir name `testdata`. As it's ignored by `go build` and commonly used by go community. Including stdlib.

smacker commented on github/downloader_test.go:
```diff
@@ -0,0 +1,205 @@
+	var teststore *store.TestStore = &store.TestStore{}
+	downloader.storer = teststore
+
+	for _, test := range tests {
```
it's a good idea to use `t.Run` for table tests usually.
also if you are tired of passing `t` everywhere `require := require.New(t)` can help with it.

lwsanty commented:
@kyrcha we also need to create a travis config in this PR

dpordomingo commented:
Also, open question for the rest of @src-d/applications 
Do we want to handle the Integration tests as in `lookout`? I mean, running them separately, i.e:
- in travis: https://github.com/src-d/lookout/blob/master/.travis.yml#L59
- in the tests itself, running it only for the passed build-tag: https://github.com/src-d/lookout/blob/master/cmd/server-test/common_test.go#L1
- in the makefile, passing the filter build-tag: https://github.com/src-d/lookout/blob/master/Makefile#L83

dpordomingo commented on github/store/teststore.go:
```diff
@@ -0,0 +1,117 @@
+	"github.com/src-d/metadata-retrieval/github/graphql"
+)
+
+type TestStore struct {
```
What if renaming this as MemoryStore, or similar.

kyrcha commented:
Thank you for your comments. To summarize:

- [ ] use `goimports`
- [ ] put data in a `testdata` dir and follow ignore rules as indicated in the official [language docs](https://golang.org/cmd/go/#hdr-Test_packages)
- [ ] refactor code to make it more modular, less wordy, more DRY and rename/restructure files and dirs
- [ ] add documentation
- [ ] add `go-log`
- [ ] use `t.Run`
- [ ] add travis config
- [ ] consider and implement proposed ways to handle the GH token

smacker commented:
btw my comments are just recommendations if you disagree with anything feel free to ignore it.

kyrcha commented:
@smacker OK. Thanks though to all, because it is compound knowledge from seasoned Golang developers.

@dpordomingo I replicated the GH token issue where you have to wait for a lot of minutes. Actually I found 4 behaviors after adding this [skip step](https://github.com/src-d/sourced-ce/blob/master/test/init_orgs_test.go#L23).

  1. If a user has not set the GITHUB_TOKEN at all: `env -u GITHUB_TOKEN go test -v ./...`
  2. If the user has a correct token according to instructions: `GITHUB_TOKEN=<correct token> go test -v ./...  `
 3. If the user has a correct token but without the necessary permissions: `GITHUB_TOKEN=<incorrect token> go test -v ./...  `
4. If the user has something else: `GITHUB_TOKEN=1234 go test -v -cover ./...  `

The results are:
 1. The test cases are skipped as expected
2. The tests *pass* in a matter of seconds
3. The tests *fail* in a matter of seconds and you get an _"organization query failed: Your token has not been granted the required scopes to execute this query"_ error
4. The tests take minutes to run and I get the same error as yours:
```
Expected nil, but got: &errors.errorString{s:"organization query failed: non-200 OK status code: 401 Unauthorized body: \"{\\\"message\\\":\\\"Bad credentials\\\",\\\"documentation_url\\\":\\\"https://developer.github.com/v4\\\"}\""}
```
So I issued the command for checking the library:

```
GITHUB_TOKEN=1234 go run examples/cmd/*.go repo --log-level=debug --version 0 --owner=kyrcha --name=metadata-retrieval
``` 
and got this:

```
[2019-10-04T15:11:29.921696+03:00] ERROR retrying in 10ms error=non-200 OK status code: 401 Unauthorized body: "{\"message\":\"Bad credentials\",\"documentation_url\":\"https://developer.github.com/v4\"}" source=github/retry.go:61
[2019-10-04T15:11:30.346296+03:00] ERROR retrying in 20ms error=non-200 OK status code: 401 Unauthorized body: "{\"message\":\"Bad credentials\",\"documentation_url\":\"https://developer.github.com/v4\"}" source=github/retry.go:61
[2019-10-04T15:11:30.786723+03:00] ERROR retrying in 60ms error=non-200 OK status code: 401 Unauthorized body: "{\"message\":\"Bad credentials\",\"documentation_url\":\"https://developer.github.com/v4\"}" source=github/retry.go:61
[2019-10-04T15:11:31.266948+03:00] ERROR retrying in 300ms error=non-200 OK status code: 401 Unauthorized body: "{\"message\":\"Bad credentials\",\"documentation_url\":\"https://developer.github.com/v4\"}" source=github/retry.go:61
[2019-10-04T15:11:32.264114+03:00] ERROR retrying in 2.7s error=non-200 OK status code: 401 Unauthorized body: "{\"message\":\"Bad credentials\",\"documentation_url\":\"https://developer.github.com/v4\"}" source=github/retry.go:61
[2019-10-04T15:11:35.389129+03:00] ERROR retrying in 10s error=non-200 OK status code: 401 Unauthorized body: "{\"message\":\"Bad credentials\",\"documentation_url\":\"https://developer.github.com/v4\"}" source=github/retry.go:61
```
which means that it retries with augmenting time steps even though the GH token is incorrect. I am guessing this is not the intended behavior. Open an issue?

kyrcha commented:
Will ask for a GH token to be installed from the infrastructure according to this [issue](https://github.com/src-d/ci/issues/51)

smacker commented on github/downloader_test.go:
```diff
@@ -2,204 +2,152 @@ package github
+	require.Equal(oracle.IsArchived, storer.Repository.IsArchived)
+	require.Equal(oracle.HasWiki, storer.Repository.HasWikiEnabled)
+	require.Equal(oracle.NumOfPrs, len(storer.Prs))
+	require.Equal(oracle.NumOfPrComments, len(storer.PrComments))
```
if you want you can use `require.Len` here.

lwsanty commented on github/downloader_test.go:
```diff
@@ -0,0 +1,153 @@
+
+func loadOnlineTests(filepath string) (OnlineTests, error) {
+	var err error
+	var tests OnlineTests
```
nitpick
```go
	var (
		err   error
		tests OnlineTests
	)
```

lwsanty commented on github/downloader_test.go:
```diff
@@ -0,0 +1,153 @@
+	checkToken(t)
+	var err error
+	tests, err := loadOnlineTests("../testdata/online-repository-tests.json")
+
```
nitpick: remove an empty line

lwsanty commented on testdata/online-organization-tests.json:
```diff
@@ -0,0 +1,14 @@
+            "numOfUsers": 0
+        }
+    ]
+}
```
nitpick: add a line break

lwsanty commented on testdata/online-repository-tests.json:
```diff
@@ -0,0 +1,101 @@
+            "numOfPrComments": 1
+        }
+    ]
+}
```
nitpick: add a line break

lwsanty commented on testutils/memory.go:
```diff
@@ -0,0 +1,120 @@
+	return nil
+}
+
+// SaveIssue TODO(Kyriakos)
```
please consider this guide on TODOs https://github.com/src-d/guide/blob/806661020bf47a2a736e7b72aee1095d42a41774/engineering/conventions.md#todo-and-fixme-comments

lwsanty commented on testutils/memory.go:
```diff
@@ -0,0 +1,120 @@
+	Topics       []string
+	Users        []*graphql.UserExtended
+	Prs          []*graphql.PullRequest
+	PrComments   []*graphql.IssueComment
```
nitpick: rename to `PRs` and `PrComments`, [reason](https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

lwsanty commented on testutils/memory.go:
```diff
@@ -0,0 +1,120 @@
+
+import (
+	"github.com/src-d/metadata-retrieval/github/graphql"
+	"gopkg.in/src-d/go-log.v1"
```
we can also split these ones  because `"github.com/src-d/metadata-retrieval/github/graphql"` belongs to this repo, but you can skip it as @smacker mentioned

dpordomingo commented on .travis.yml:
```diff
@@ -0,0 +1,17 @@
+  - 1.13.x
+env:
+  global:
+    - SOURCED_GITHUB_TOKEN=$GITHUB_TOKEN
```
Why're using `GITHUB_TOKEN` for the env var storing the github token
https://github.com/src-d/metadata-retrieval/pull/22/files#diff-fc4f49f574c8bbba1fbc6c64589a5536R49

kyrcha commented:
@dpordomingo @lwsanty @smacker 

This is not the final implementation resolving #13. Will start working on the recording/replaying right after this.

Ready for last sanity check:
- Removed unnecessary env variable declaration from travis
- Removed Carlos' commits 
- Rebased against master using this [guide](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request)
- Fixed formatting errors

dpordomingo commented on github/downloader_test.go:
```diff
@@ -0,0 +1,153 @@
+}
+
+func checkToken(t *testing.T) {
+	if os.Getenv("GITHUB_TOKEN") == "" {
```
From the PR desc, I'd say that the purpose of these tests, is to ensure that downloader can fetch an org and its repositories with its descendants, so having a `GITHUB_TOKEN` could be considered a hard requirement to pass the tests.

If so, I'd require such token instead of skipping the tests (which is not detected in Travis, and can only seen running the tests with `-v` flag.

```diff
 func checkToken(t *testing.T) {
-       if os.Getenv("GITHUB_TOKEN") == "" {
-               t.Skip("GITHUB_TOKEN is not set")
-               return
-       }
+       req := require.New(t)
+       req.NotEmpty(os.Getenv("GITHUB_TOKEN"), "GITHUB_TOKEN is not set")
 }
```

kyrcha replied:
Probably I misunderstood since I too agree that the tests should fail if there is no token. I got sidetracked because travis does not transfer env vars when you push an updated from a fork, so all PRs from a fork will fail. In any case I will change this.

dpordomingo replied:
I didn't realize that detail, that will explain why integration tests requiring `GITHUB_TOKEN` are skipped in `sourced-ce` PR's (example: https://travis-ci.com/src-d/sourced-ce/jobs/240414245#L377), but they run on `master` (e.g. https://travis-ci.com/src-d/sourced-ce/jobs/240651163#L376).

And also why `ghsync` tests are using a custom `http.Client` [without authentication](https://github.com/src-d/ghsync/blob/master/utils/rate_test.go#L103), instead of the regular one, [with authentication](https://github.com/src-d/ghsync/blob/master/cmd/ghsync/subcmd/common.go#L132); approach that I think we can not take here because GH GraphQL API v4 requires authentication.

Considering that, I wonder if it'd be better if we replicate `source-ce` behavior about skipping the tests requiring `GITHUB_TOKEN` (in order to avoid having failed builds in PRs) and discuss an alternative in a new issue with the rest of apps and qa team.

kyrcha replied:
OK, I'll wait till later today and then I will skip the tests that require the token since I will another PR with tests that do not require it. 
So we can replicate the `source-ce` behavior where the master will run both online and offline tests, but in PRs forks will run only the offline ones based on http "recordings".

lwsanty replied:
@dpordomingo `GITHUB_TOKEN` in PRs is not only an apps team pain, I'm pretty sure that each team has bumped into it at least once and probably we should discuss it on the higher level than the team's one

As far as testing team has an editors access I suggest @kyrcha to create further PRs from origin branches

Anyways PRs from forks are still an open question

kyrcha replied:
I have one more PR ready to push from my fork and then I will start PRs from origin

se7entyse7en replied:
I wasn't aware of this behavior of env vars, but it totally makes sense.

 > I'm pretty sure that each team has bumped into it at least once and probably we should discuss it on the higher level than the team's one.

yes please! 👍

dpordomingo commented:
Travis failing because of [a missing `GITHUB_TOKEN` as explained above](https://github.com/src-d/metadata-retrieval/pull/22#discussion_r332245564).

What do you think would be better?
- to have failing builds in PRs and running ones in `master`?
- skip tests that would fail in PRs?
- other?

I'd choose the first, as in `sourced-ce` for simplicity, and consider an alternative in a separate issue (like handle it from Travis, so tests will fail if they're run locally or in `src-d/master`, but they will be skipped if run in PRs)

dpordomingo commented:
Also, I think we should create an issue like [this one](https://github.com/src-d/infrastructure/issues/1154), to ask for `GITHUB_TOKEN` secret in Travis, shouldn't we @se7entyse7en ?

lwsanty commented on testutils/memory.go:
```diff
@@ -0,0 +1,123 @@
+
+// SaveIssueComment noop
+func (s *Memory) SaveIssueComment(repositoryOwner, repositoryName string, issueNumber int, comment *graphql.IssueComment) error {
+	log.Infof("  issue comment data fetched by %s at %v: %q\n", comment.Author.Login, comment.CreatedAt, trim(comment.Body))
```
why do we need these spaces?

kyrcha replied:
I just replicated the [stdout](https://github.com/src-d/metadata-retrieval/blob/master/github/store/stdout.go) store where spaces are used for identing hierarchical structures (Issue > Issue Comments etc.)

lwsanty replied:
will it make sense to add `\t` instead of spaces, wdyt?

kyrcha replied:
Yes, will add them to be more elegant

lwsanty commented on .travis.yml:
```diff
@@ -0,0 +1,14 @@
+language: go
```
@src-d/applications data-retrieval team has a separate job for osx https://github.com/src-d/gitcollector/blob/5d60f873b2a9c74de7cc4544692ea3d85d09092e/.travis.yml#L37 
Is it optional?

smacker replied:
we also usually test on macOs separately. And Windows as well.

lwsanty replied:
@kyrcha let's do it in a separate PR

kyrcha replied:
@lwsanty Just pushed

kyrcha replied:
I am removing windows since it does not work out of the box

se7entyse7en replied:
It would be great if you could open an issue 🙏

se7entyse7en commented on README.md:
```diff
@@ -359,3 +359,21 @@ You can see the diff between the current DB schema and the ghsync schema here:
+
+for coverage information.
+
+Where `GITHUB_TOKEN` is a personal access token as described above.
```
Could you just add the scope required by the token please?

kyrcha replied:
Added
